### ISSUE TITLE
Long pause between 2 training epochs. (A training epoch = 6m40s, while pause = 4m20s)

### ISSUE THREAD

**Issue by shen-ttt** [visionml/pytracking#374]
I have been testing the training pipeline of ToMP 50 and noticed a big pause between 2 epochs. 

For 8 GPUs, the training time for 1 epoch is around 6:40 and the pause takes another 4:20, which is not quite efficient. 

I checked the ltr_trainer code and seems like everything is pretty lightweight.

Could give me a clue about what's going on here? Anything we can do it optimize? 

Thank you!

**Comment by 2006pmach** [comment]
This has something to do with the DataLoader and how many GPUs you use. I never had such a large overhead but it on your hardware details and how many GPUs you use. If training one epoch gets faster the relative overhead is larger. The problem is that after every epoch the worker processes containing the datasets are closed and need to be reopened which takes time for all the datasets we use. 

An easy fix would be to reduce the number of epochs and increase the number of randomly sampled training samples per epoch let say by a factor of 10 (from 200 epochs to 20 and from 40k samples per epoch to 400k samples). This should be fine since we anyway sample randomly instead of iterating over the whole dataset. 

Another solution would be to avoid that the working processes are closed. Here we can exploit the fact that we sample randomly from all datasets. You can add the following to `ltr/data/loader.py`

```python
class MultiEpochLTRLoader(LTRLoader):

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._DataLoader__initialized = False
        self.batch_sampler = _RepeatSampler(self.batch_sampler)
        self._DataLoader__initialized = True
        self.iterator = super().__iter__()

    def __len__(self):
        return len(self.batch_sampler.sampler)

    def __iter__(self):
        for i in range(len(self)):
            yield next(self.iterator)


class _RepeatSampler(object):
    """ Sampler that repeats forever.
    Args:
        sampler (Sampler)
    """

    def __init__(self, sampler):
        self.sampler = sampler

    def __iter__(self):
        while True:
            yield from iter(self.sampler)
```

Now instead of calling `loader_train = LTR('train', dataset_train, training-True, ...)` use ``loader_train = MultiEpochLTR('train', dataset_train, training-True, ...)` in the `train_settings` files.

**Comment by shen-ttt** [comment]
Thank you for the detailed explanation! Right now I am reducing the amount of epochs and increase the sampling quota of each epoch and it works fine. Will defintely try to use your approach for a more elegant solution. 

Also, will close this issue now.

**Comment by NewboySoCool** [comment]
Nice， but I have a question about MultiEpochLTRLoader. 
 Does this implementation messes up the shuffling in Dataloader? I always thought the Dataloader reinitializes itself for the shuffling and a couple other operations.
Are there's no any drawbacks from using this method? It's very weird PyTorch devs never considered doing such a crucial thing.

And it's suitable for DDP？ and the sampler.set_epoch(epoch)?
